File: frontend/src/stores/tags.ts
import { defineStore } from 'pinia'
import { ref } from 'vue'
import { tagsApi } from '@/api/tags'
import type { Tag, TagCreate, TagUpdate } from '@/types'

export const useTagsStore = defineStore('tags', () => {
  const tags = ref<Tag[]>([])
  const loading = ref(false)

  async function fetchTags() {
    loading.value = true
    try {
      const response = await tagsApi.list()
      tags.value = response.data
    } finally {
      loading.value = false
    }
  }

  async function createTag(data: TagCreate) {
    try {
      const response = await tagsApi.create(data)
      tags.value.push(response.data)
      return response.data
    } catch (error: any) {
      const message = error?.response?.data?.detail || '创建标签失败'
      throw new Error(message)
    }
  }

  async function updateTag(id: string, data: TagUpdate) {
    try {
      const response = await tagsApi.update(id, data)
      const index = tags.value.findIndex(t => t.id === id)
      if (index !== -1) {
        tags.value[index] = response.data
      }
      return response.data
    } catch (error: any) {
      const message = error?.response?.data?.detail || '更新标签失败'
      throw new Error(message)
    }
  }

  async function deleteTag(id: string) {
    try {
      await tagsApi.delete(id)
      tags.value = tags.value.filter(t => t.id !== id)
    } catch (error: any) {
      const message = error?.response?.data?.detail || '删除标签失败'
      throw new Error(message)
    }
  }

  return {
    tags,
    loading,
    fetchTags,
    createTag,
    updateTag,
    deleteTag
  }
})

File: backend/app/api/routes/tags.py
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session
from typing import List
from app.db.session import get_db
from app.models.tag import Tag
from app.models.note import Note
from app.models.user import User
from app.schemas import Response, TagCreate, TagUpdate, TagResponse
from app.api.deps import get_current_user

router = APIRouter(prefix="/tags", tags=["标签"])

@router.get("", response_model=Response[List[TagResponse]])
async def get_tags(
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    tags = db.query(Tag).all()
    result = []
    for tag in tags:
        note_count = db.query(Note).join(Note.tags).filter(Tag.id == tag.id).count()
        result.append(TagResponse(
            id=tag.id,
            name=tag.name,
            color=tag.color,
            note_count=note_count,
            created_at=tag.created_at
        ))
    return Response(data=result)

@router.post("", response_model=Response[TagResponse])
async def create_tag(
    tag_data: TagCreate,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    existing = db.query(Tag).filter(Tag.name == tag_data.name).first()
    if existing:
        raise HTTPException(status_code=400, detail="标签名称已存在")
    
    tag = Tag(
        name=tag_data.name,
        color=tag_data.color
    )
    db.add(tag)
    db.commit()
    db.refresh(tag)
    
    return Response(data=TagResponse(
        id=tag.id,
        name=tag.name,
        color=tag.color,
        note_count=0,
        created_at=tag.created_at
    ), message="标签创建成功")

@router.put("/{tag_id}", response_model=Response[TagResponse])
async def update_tag(
    tag_id: str,
    tag_data: TagUpdate,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    tag = db.query(Tag).filter(Tag.id == tag_id).first()
    if not tag:
        raise HTTPException(status_code=404, detail="标签不存在")
    
    if tag_data.name is not None:
        existing = db.query(Tag).filter(Tag.name == tag_data.name, Tag.id != tag_id).first()
        if existing:
            raise HTTPException(status_code=400, detail="标签名称已存在")
        tag.name = tag_data.name
    if tag_data.color is not None:
        tag.color = tag_data.color
    
    db.commit()
    db.refresh(tag)
    
    note_count = db.query(Note).join(Note.tags).filter(Tag.id == tag.id).count()
    
    return Response(data=TagResponse(
        id=tag.id,
        name=tag.name,
        color=tag.color,
        note_count=note_count,
        created_at=tag.created_at
    ), message="标签更新成功")

@router.delete("/{tag_id}")
async def delete_tag(
    tag_id: str,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    tag = db.query(Tag).filter(Tag.id == tag_id).first()
    if not tag:
        raise HTTPException(status_code=404, detail="标签不存在")
    
    db.delete(tag)
    db.commit()
    return Response(message="标签删除成功")
